# Bioconductor Classes Quick Overview

## The GRanges Class

```{r message=FALSE}
library(GenomicRanges)
library("Gviz")
library("DESeq2")
```

Create a GRanges object

```{r}
(A1 <- GRanges(seqnames = Rle("ch1"),
              ranges   = IRanges(start = c(10, 22, 40),
                                 width = 5),
              strand   = "+"))
```

Plot the ranges

```{r}

options(ucscChromosomeNames=FALSE)
geTrack <- GenomeAxisTrack()
deTrack <- AnnotationTrack(range =  A1)
plotTracks(c(geTrack,deTrack),from=1, to=50)
```

Show all methods for the GRanges class

```{r}
methods(class="GRanges")
```

### Intra-range methods (transform ranges independently)

**shift(x)**: moves left/right

**promoters(x)**: returns promoter ranges extending around the transcription start site (TSS) defined as start(x)

### Inter-range methods (transform all the ranges together) 

**reduce(x)**:  orders the ranges in x from left to right, then merges the overlapping or adjacent ones

**disjoin(x)**: breaks into discrete ranges based on original starts/ends



For example:

```{r}
 length(A1)
 width(A1)
 seqnames(A1)
 ranges(A1)
 promoters(A1,upstream=10, downstream=15)
```

Add metdata to the GRanges object

```{r}
 names(A1) <- LETTERS[1:3]
 genome(A1) <- "hg19"
 mcols(A1)$bioreplicate <-  "I"
 A1
```
 
Coerce the GRange into a dataframe object
 
```{r}
as.data.frame(A1)
```

### Multiple GRange Objects 

Create two more objects

```{r}
A2 <- GRanges(seqnames = Rle("ch1"),
              ranges   = IRanges(start = c(8, 22, 29),
                                 width = 5),
              strand   = "+")

 names(A2) <- LETTERS[1:3]
 genome(A2) <- "hg19"
 mcols(A2)$bioreplicate <-  "II"
 
A3 <- GRanges(seqnames = Rle("ch1"),
              ranges   = IRanges(start = c(6, 15, 36),
                                 width = 5),
              strand   = "+")

 names(A3) <- LETTERS[1:3]
 genome(A3) <- "hg19"
 mcols(A3)$bioreplicate <-  "III"
 
 A2
 A3
```

Combine the GRange objects
```{r}
(A <-  c(A1, A2, A3))
```

Merge
```{r}
(A_merged <- reduce(A))
```

Plot it
```{r}
deTrack <- AnnotationTrack(range =  A, group = A$bioreplicate)
mrgTrack <- AnnotationTrack(range =  A_merged, name = "Merged")
plotTracks(c(geTrack,deTrack, mrgTrack),groupAnnotation = "group",from=1, to=50)
```

 
To find ranges that repeat we can count the number of overlapping regions and remove merged segments that have a certain number of counts (e.g. 1)
```{r}
(cnts <- countOverlaps(A_merged,A))
A_merged_consensus <- A_merged[-which(cnts==1),]
```

```{r}
mrg_con_Track <- AnnotationTrack(range =  A_merged_consensus,name = "Merged Consensus")
plotTracks(c(geTrack, deTrack, mrgTrack, mrg_con_Track),groupAnnotation = "group",from=1, to=50)
```


## GRangesList Class

The GRangesList class is a container for a list of GRanges objects that have the same metadata column variables. First we create another GRanges object.

```{r}
(B <- GRanges(seqnames = Rle("ch1"),
              ranges   = IRanges(start = c(1,3,10,13,24,30),
                                 width = 5),
              strand   = "+"))

 names(B) <- rep(LETTERS[1:3],2)
 genome(B) <- "hg19"
 mcols(B)$bioreplicate <-  rep(c("I","II"),c(3,3))
# combine
AB <- GRangesList(A, B)
names(AB) <- c("A", "B")
(AB)
```

we can reduce each object independently  
```{r}
(AB_merged <- reduce(AB))
```

The GRangesList class is a List object (not list) of GenomicRanges 

```{r}
(ABg <- GenomicRangesList(A, B))
```

We can also reduce the GRanges objects of all conditions into one GRange

```{r}

(AB_combined <- unlist(GenomicRangesList(A, B)))
(AB_combined_merged <- reduce(AB_combined ))

```

##  RangedSummarizedExperiment


RangedSummarizedExperiment is a array-like container in which rows represent features of interest (e.g. genes) and columns represent samples.


![](RangedSummarizedExperiment.png)


First we list the accessor functions for the class

```{r}
methods(class="RangedSummarizedExperiment")
```

We can construct a RangedSummarizedExperiment using the class constructor *SummarizedExperiment(assays, rowRanges, colData)*  or it can be generated by other functions such as the *summarizeOverlaps* function that extends *findOverlaps* by providing options to resolve reads that overlap multiple features. For example given a list of bam files, we can geenerate a RangedSummarizedExperiment as such

```{r, eval=FALSE}
  se <- summarizeOverlaps(
    features = A_merged_consensus,
    reads = bam_files,
    mode = "Union",
    singleEnd = FALSE,
    ignore.strand = TRUE,
    fragments = TRUE,
    inter.feature = FALSE
  )

assays(se)[[1]]
colData(se)
colnames(se)

```



## DESeqDataSet Class
DESeqDataSet is a subclass of RangedSummarizedExperiment that serves as a container for differential expression analysis results.  The *DESeqDataSet(se, design)* class constructor takes non-negative integer values in the counts *se* matrix and a formula to specifies the design of the experiment.


First we list the accessor functions for the class

```{r}
methods(class="DESeqDataSet")
```

Examples of accessor functions for the class

```{r, eval=FALSE}
  DESeq_dataset <- DESeqDataSet(se, design = ~ biosample)
  DESeq_dataset <-DESeq_dataset[rowSums(counts(DESeq_dataset)) > 1, ]
  rld <- rlog(DESeq_dataset, blind=FALSE)
  #fragments per kilobase per million mapped fragments
  data_rpkm <-fpkm(rld) 
  sampleDists <- dist( t( assay(rld) ) )
  ddsc  <- DESeq(DESeq_dataset)
  res <- results(dds)
  plotMA(res, ylim=c(-10,10))
```

To conduct differential analysis 

```{r, eval=FALSE}
  ddsc  <- DESeq(DESeq_dataset)
 #DESeqResults object
  res <- results(dds)
  plotMA(res, ylim=c(-10,10))
  summary(res)
```

